### ISSUE TITLE
Error between attacks 

### ISSUE THREAD

**Issue by lennyhans** [the3dfxdude/7kaa#200]
Hi, I was playing and then get an error message from OUNITAT2.cpp
```
// ------- reduce power --------//
cur_power -= attackInfo->consume_power;
err_when(cur_power < 0);
//### begin alex 28/10 ###//
if(cur_power<0) // ***** BUGHERE
  cur_power = 0;
//#### end alex 28/10 ####//
```
apparently was expected, I have not investigated more yet

**Comment by the3dfxdude** [comment]
Are you using the debug flag? Do you have a save file?

**Comment by lennyhans** [comment]
Yeah I have the file, about the debug flag I think I was not using it, but I will have to check.

[Error.sav](https://gofile.io/d/2YZ6ib)

**Comment by the3dfxdude** [comment]
Unfortunately it is telling me today that the file does not exist.

I have tried catching the error on my own running a few simulations, and so far, no luck.

**Comment by lennyhans** [comment]
Let's try again ! [Error.sav](https://file.io/2iUeZPf55LW2)

**Comment by the3dfxdude** [comment]
That's it. Looks like an assertion. I see it now. Now to find what could have triggered it.

**Comment by the3dfxdude** [comment]
I've tried simulating, and looked over the code. I can't see any way to trigger it. How did you compile this? What options were used?

**Comment by the3dfxdude** [comment]
Well, it's worse than just that assertion. The assertion is tripped, and you had the debug flag on, which saved the game at that point. The reason why it's tripped is looking like some kind of memory corruption occurred just before then. The unit in question is somehow at coordinates 832,3104, which is impossible. I checked this at the moment the game is loaded, so the save file is corrupted too.

I won't be able to debug the exact reason as there isn't enough information to trace back to the original problem. If you don't have any other special compiler options on, then I suppose then there is a legitimate issue somewhere in the Unit code? But it's pretty rare. I have a pretty good idea what to do to root out these problems by adding some checks in the code. The game has some unsafe handling of arrays through nested calls that need to be weeded out. See the issue fixed on 2019-7-29, and the SpyArrayLock.

I might go ahead and add UnitArrayLocks on functions and close this on that.

**Comment by the3dfxdude** [comment]
So I had this thought that UnitArray would have the same problem as SpyArray, because they are similar right? No they are not. UnitArray uses DynArray::get_ptr and SpyArray uses DynArray::get. And by UnitArray storing a copy of a pointer somewhere else, and the pointer being based by body_buf in SpyArray, UnitArray does not have this issue.

I checked all the arrays, and I think the only array that really has an issue is SpyArray. Since most of the complex code uses get_ptr, I think I'll just convert SpyArray over. But this doesn't help this issue. So unless there is another piece of info, I eventually close this.

**Comment by lennyhans** [comment]
Sorry for the delay, for now I do not have any additional info to point out the problem, but as far as I  remember I was playing with meson build system while this happen, so I will to try to compile it the same way with autoconf or at least to get the compile variable values to see if there are anything relevant. Thanks to give it that deep evaluation!

**Comment by the3dfxdude** [comment]
Ok so I have been able to work on quite a bit, and got a better picture. It does not look like a memory corruption. The save is probably ok. So the question is what happened. Well I just got lucky today. I had been running a number of simulations and I finally hit it. It has something to do with egyptian unit's attack. I'll dissect the new case when I get time.

**Comment by the3dfxdude** [comment]
Closing this on 03acb27490bd41b3a29d97e9017faee8470c1a5f

This does appear to be a bug. Understanding it and choosing a fix was not straight forward. I would hope for a cleaner way to handle this, and 7k2 attempted to address some related things, and structure it better. But it is still quite complex and changing this thing at too high a level could affect alot, a poses not exactly related questions on what the intended gameplay should be. I opted for the narrowest fix, only address the thing in the moment things really start going wrong, which later trigger the assertion a few frames later, which puts the function into compliance.

I did learn some things about this game, how much detail they are simulating here. It's amazing they tried these things in '97... Not sure if anyone else has done this! Even if the code isn't fully fleshed out.

I don't think the extra flame arrow attack was a thing that can be easily exploited, and the AI almost certainly never find itself in the required scenario. There may be some other related problems, which will need to be addressed by improving code, but they are really hard to find, I think. So it's hard to say that the Egyptians are overpowered. In review, I think we should practice these various match-ups between races again, because there is alot of game detail that can be legitimately exploited, by design, that could turn the course of a battle, that no one ever really things about :)
